src/owners_file.rs

```diff
@@ -189,16 +216,57 @@ fn resolve_include_path<P0: AsRef<Path>, P1: AsRef<Path>, P2: AsRef<Path>>(
         )
     })?;
 
-    if include_path_ref.is_absolute() {
-        return Ok(repo_base_path.join(
+    let path = if include_path_ref.is_absolute() {
+        repo_base_path.join(
             include_path_ref
                 .strip_prefix("/")
                 .or_else(|_| include_path_ref.strip_prefix("\\"))
                 .unwrap_or(include_path_ref),
-        ));
+        )
     } else {
-        return Ok(current_dir.join(include_path_ref));
+        current_dir.join(include_path_ref)
     };
+
+    let canonicalized_path = fs::canonicalize(&path).map_err(|error| {
+        anyhow!(
+            "Failed to canonicalize include path '{}': {}",
+            path.display(),
+            error
+        )
+    })?;
+    if !canonicalized_path.starts_with(repo_base_path) {
+        return Err(anyhow!(
+            "Include path '{}' is outside the repository base '{}'.",
+            canonicalized_path.display(),
+            repo_base_path.display()
+        ));
+    }
+    Ok(canonicalized_path)
+}
+
+fn format_include_cycle_error(
+    path: &PathBuf,
+    seen_owners: &HashMap<PathBuf, Option<PathBuf>>,
+) -> anyhow::Error {
+    let mut chain = vec![path.clone()];
+    let mut current = path;
+
+    // Walk the reverse linked list to get a nice printable chain of includes.
+    while let Some(Some(parent)) = seen_owners.get(current) {
+        chain.push(parent.clone());
+        current = parent;
+    }
+
+    // The error message is easier to read if it's in forward order rather than reverse.
+    chain.reverse();
+
+    let message = chain
+        .iter()
+        .map(|p| p.display().to_string())
+        .collect::<Vec<_>>()
+        .join(" -> ");
+
+    return anyhow!("Cycle detected in includes: {}", message);
```

Comment:
[optional] The output might be easier to parse if we put it on multiple lines, similar to a stack trace.
```suggestion
        .join("\n     -> ");

    return anyhow!("Cycle detected in includes:\n.   {}", message);
```